planner/core/optimizer.go

```diff
@@ -144,6 +144,16 @@ func postOptimize(sctx sessionctx.Context, plan PhysicalPlan) PhysicalPlan {
 }
 
 func logicalOptimize(ctx context.Context, flag uint64, logic LogicalPlan) (LogicalPlan, error) {
+	// if there is something after flagPrunColumns, do flagPrunColumnsAgain
+	if flag&flagPrunColumns > 0 && flag-flagPrunColumns > flagPrunColumns {
+		flag |= flagPrunColumnsAgain
+	}
+	return logicalOptimize4Test(ctx, flag, logic)
+}
+
+// logicalOptimize4Test is separated for test because flagPrunColumnsAgain
+// in logicalOptimize may reset all keys in Schema.
+func logicalOptimize4Test(ctx context.Context, flag uint64, logic LogicalPlan) (LogicalPlan, error) {
```

Comment:
How about the name `doLogicalOptimize`?

Reply:
Prefer `logicalOptimize4Test` because it reminds users only calling it in tests.